In [ ]:
# Q1
import pandas as pd

df = pd.read_csv("sparklingwine.csv")
print(df.head())

   Unnamed: 0  fixed acidity  volatile acidity  citric acid  residual sugar  \
0           0            5.4              0.29         0.38             1.2   
1           1            6.7              0.24         0.29            14.9   
2           2            6.8              0.33         0.31             7.4   
3           3            6.4              0.27         0.19             2.0   
4           4            6.1              0.30         0.30             2.1   

   chlorides  free sulfur dioxide  total sulfur dioxide  density    pH  \
0      0.029                 31.0                 132.0  0.98895  3.28   
1      0.053                 55.0                 136.0  0.99839  3.03   
2      0.045                 34.0                 143.0  0.99226  3.06   
3      0.084                 21.0                 191.0  0.99516  3.49   
4      0.031                 50.0                 163.0  0.98950  3.39   

   sulphates  alcohol  quality  
0       0.36     12.4        6  
1       0.52  

In [ ]:
#Q2

df["good wine"] = (df["quality"] >= 6).astype(int)

df[["quality", "good wine"]].head()

# df = df.drop(columns=["Unnamed: 0"]) # remove Unnamed: 0 column (not a feature)
 


quality  good wine
0        6          1
1        5          0
2        6          1
3        4          0
4        7          1

In [ ]:
#Q3

train_df = df.iloc[:900]
val_df = df.iloc[900:1200]
test_df = df.iloc[1200:1600]

print(len(train_df), len(val_df), len(test_df))


900 300 400


In [7]:
#Q4
X_train = train_df.drop(columns=["good wine"])
y_train = train_df["good wine"]

X_val = val_df.drop(columns=["good wine"])
y_val = val_df["good wine"]

X_test = test_df.drop(columns=["good wine"])
y_test = test_df["good wine"]

mu = X_train.mean()
sigma = X_train.std()

# normalisation
X_train_z = (X_train - mu) / sigma
X_val_z   = (X_val   - mu) / sigma
X_test_z  = (X_test  - mu) / sigma

#print(X_train_z.mean())
#print(X_train_z.std())

print(X_train_z.mean().round(2))
print(X_train_z.std().round(2))

Unnamed: 0              0.0
fixed acidity          -0.0
volatile acidity        0.0
citric acid             0.0
residual sugar         -0.0
chlorides              -0.0
free sulfur dioxide    -0.0
total sulfur dioxide    0.0
density                -0.0
pH                     -0.0
sulphates              -0.0
alcohol                -0.0
quality                 0.0
dtype: float64
Unnamed: 0              1.0
fixed acidity           1.0
volatile acidity        1.0
citric acid             1.0
residual sugar          1.0
chlorides               1.0
free sulfur dioxide     1.0
total sulfur dioxide    1.0
density                 1.0
pH                      1.0
sulphates               1.0
alcohol                 1.0
quality                 1.0
dtype: float64


In [ ]:
# Q5
from sklearn.neighbors import KNeighborsClassifier

# train kNN models for k=1 to 100
knn_models = {} # sore in dictionary

for k in range(1, 101):
    knn = KNeighborsClassifier(n_neighbors=k)
    knn.fit(X_train_z, y_train) # train classifier on the normalised training data
    knn_models[k] = knn # store trained model in dictionary



In [12]:
# Q6
# evaluate in validation data
val_accuracy = {} # dictionary to store classification accuracy of validation set

for k in range(1, 101):
    knn = KNeighborsClassifier(n_neighbors=k)
    knn.fit(X_train_z, y_train) # train the model using the normalised training data
    val_accuracy[k] = knn.score(X_val_z, y_val) # return mean classification accuracy w score()

# best classifier
best_k = max(val_accuracy, key=val_accuracy.get) # max validation accuracy
best_k, val_accuracy[best_k] 


(9, 0.9133333333333333)

In [13]:
# Q7

# train using training and validation data
# Combine training and validation data
X_trainval_z = pd.concat([X_train_z, X_val_z])
y_trainval = pd.concat([y_train, y_val])

# Train final kNN classifier with the chosen k
final_knn = KNeighborsClassifier(n_neighbors=best_k)
final_knn.fit(X_trainval_z, y_trainval)


# predict on test data
y_test_pred = final_knn.predict(X_test_z)


# test accuracy
test_accuracy = final_knn.score(X_test_z, y_test)
test_accuracy


# convert accuracy to generalisation error
generalisation_error = 1 - test_accuracy
generalisation_error




0.07499999999999996

In [15]:
# Q8
train_df2 = df.iloc[:400]
val_df2 = df.iloc[400:800]
test_df2 = df.iloc[800:1600]

X_train2 = train_df2.drop(columns=["good wine"])
y_train2 = train_df2["good wine"]

X_val2 = val_df2.drop(columns=["good wine"])
y_val2 = val_df2["good wine"]

X_test2 = test_df2.drop(columns=["good wine"])
y_test2 = test_df2["good wine"]

mu2 = X_train2.mean()
sigma2 = X_train2.std()

# normalisation
X_train_z2 = (X_train2 - mu2) / sigma2
X_val_z2 = (X_val2  - mu2) / sigma2
X_test_z2  = (X_test2  - mu2) / sigma2

print(X_train_z2.mean().round(2))
print(X_train_z2.std().round(2))

# train kNN models for k=1 to 100
knn_models2 = {} # sore in dictionary

for k in range(1, 101):
    knn2 = KNeighborsClassifier(n_neighbors=k)
    knn2.fit(X_train_z2, y_train2) # train classifier on the normalised training data
    knn_models2[k] = knn2 # store trained model in dictionary


val_accuracy2 = {} # dictionary to store classification accuracy of validation set

for k in range(1, 101):
    knn2 = KNeighborsClassifier(n_neighbors=k)
    knn2.fit(X_train_z2, y_train2) # train the model using the normalised training data
    val_accuracy2[k] = knn2.score(X_val_z2, y_val2) # return mean classification accuracy w score()

# best classifier
best_k2 = max(val_accuracy2, key=val_accuracy2.get) # max validation accuracy
best_k2, val_accuracy2[best_k2] 

# train using training and validation data
# Combine training and validation data
X_trainval_z2 = pd.concat([X_train_z2, X_val_z2])
y_trainval2 = pd.concat([y_train2, y_val2])

# Train final kNN classifier with the chosen k
final_knn2 = KNeighborsClassifier(n_neighbors=best_k2)
final_knn2.fit(X_trainval_z2, y_trainval2)


# predict on test data
y_test_pred2 = final_knn2.predict(X_test_z2)


# test accuracy
test_accuracy2 = final_knn2.score(X_test_z2, y_test2)
test_accuracy2


# convert accuracy to generalisation error
generalisation_error2 = 1 - test_accuracy2
generalisation_error2


Unnamed: 0             -0.0
fixed acidity          -0.0
volatile acidity       -0.0
citric acid            -0.0
residual sugar         -0.0
chlorides              -0.0
free sulfur dioxide     0.0
total sulfur dioxide    0.0
density                -0.0
pH                     -0.0
sulphates               0.0
alcohol                 0.0
quality                -0.0
dtype: float64
Unnamed: 0              1.0
fixed acidity           1.0
volatile acidity        1.0
citric acid             1.0
residual sugar          1.0
chlorides               1.0
free sulfur dioxide     1.0
total sulfur dioxide    1.0
density                 1.0
pH                      1.0
sulphates               1.0
alcohol                 1.0
quality                 1.0
dtype: float64


0.14500000000000002

0.07499999999999996 vs 0.14500000000000002

The generalisation error increases from approximately 0.075 to 0.145 when the training set size is reduced, indicating worse performance with less training data.

This sensitivity reflects the data-driven nature of kNN, which relies heavily on having many nearby training points to make accurate predictions.

With more training data (first split), the classifier generalises better and produces more reliable predictions on unseen data.

With fewer training samples (second split), the classifier struggles to capture the underlying structure of the data, leading to higher test error.

Overall, the classifier is reasonably well-suited to the dataset when sufficient training data are available, but its performance degrades noticeably when the training set is smaller.

---

Using the first data split (900 training, 300 validation, 400 test), the estimated generalisation error is approximately 0.075, whereas under the second split (400 training, 400 validation, 800 test) the generalisation error increases to approximately 0.145. This difference is substantial and suggests that the classifier’s performance is sensitive to the amount of training data available. With fewer training samples, the kNN classifier is less able to accurately capture the underlying structure of the data, leading to poorer generalisation on unseen observations. This behaviour is expected for kNN, which is a non-parametric, data-driven method whose performance improves as the size of the training set increases.

The classifier can be judged as reasonably well-suited to the data set in the first split, where the larger training set leads to a relatively low test error and stable performance. However, the marked increase in generalisation error under the second split suggests that the classifier relies heavily on having sufficient training data and may not generalise as reliably when data are scarce. Overall, this indicates that while kNN can perform well for this problem, its effectiveness depends strongly on the size of the training set, and care must be taken when deploying it in settings with limited labelled data.